frontend/src/metabase/lib/visualization_settings.js

```diff
@@ -1,230 +1,573 @@
-        'zoom': 10,
-        'center_latitude': 37.7577, //defaults to SF ;-)
-        'center_longitude': -122.4376
-    }
-};
+import { getCardColors, getFriendlyName } from "metabase/visualizations/lib/utils";
 
-var groupsForVisualizations = {
-    'scalar': ['global'],
-    'table': ['global', 'columns'],
-    'pie': ['global', 'chart', 'pie'],
-    'bar': ['global', 'columns', 'chart', 'xAxis', 'yAxis', 'bar'],
-    'line': ['global', 'columns', 'chart', 'xAxis', 'yAxis', 'line'],
-    'area': ['global', 'columns', 'chart', 'xAxis', 'yAxis', 'line', 'area'],
-    'country': ['global', 'columns', 'chart', 'map'],
-    'state': ['global', 'columns', 'chart', 'map'],
-    'pin_map': ['global', 'columns', 'chart', 'map']
-};
+import ChartSettingInput from "metabase/visualizations/components/settings/ChartSettingInput.jsx";
+import ChartSettingInputNumeric from "metabase/visualizations/components/settings/ChartSettingInputNumeric.jsx";
+import ChartSettingSelect from "metabase/visualizations/components/settings/ChartSettingSelect.jsx";
+import ChartSettingToggle from "metabase/visualizations/components/settings/ChartSettingToggle.jsx";
+import ChartSettingFieldsPicker from "metabase/visualizations/components/settings/ChartSettingFieldsPicker.jsx";
+import ChartSettingColorsPicker from "metabase/visualizations/components/settings/ChartSettingColorsPicker.jsx";
+import ChartSettingOrderedFields from "metabase/visualizations/components/settings/ChartSettingOrderedFields.jsx";
 
-export function getDefaultColor() {
-    return DEFAULT_COLOR;
+function columnsAreValid(colNames, data, filter = () => true) {
+    if (typeof colNames === "string") {
+        colNames = [colNames]
+    }
+    if (!data || !Array.isArray(colNames)) {
+        return false;
+    }
+    const colsByName = {};
+    for (const col of data.cols) {
+        colsByName[col.name] = col;
+    }
+    return colNames.reduce((acc, name) =>
+        acc && (name == undefined || (colsByName[name] && filter(colsByName[name])))
+    , true);
 }
 
-export function getDefaultColorHarmony() {
-    return DEFAULT_COLOR_HARMONY;
+function getSeriesTitles([{ data: { rows, cols } }], vizSettings) {
+    const seriesDimension = vizSettings["graph.dimensions"][1];
+    if (seriesDimension != null) {
+        let seriesIndex = _.findIndex(cols, (col) => col.name === seriesDimension);
+        return crossfilter(rows).dimension(d => d[seriesIndex]).group().all().map(v => v.key);
+    } else {
+        return vizSettings["graph.metrics"].map(name => {
+            let col = _.findWhere(cols, { name });
+            return col && getFriendlyName(col);
+        });
+    }
 }
 
-function getSettingsForGroup(dbSettings, groupName) {
-    if (typeof dbSettings != "object") {
-        dbSettings = {};
+function getDefaultDimensionsAndMetrics([{ data: { cols, rows } }]) {
+    let type = getChartTypeFromData(cols, rows, false);
+    switch (type) {
+        case DIMENSION_DIMENSION_METRIC:
```

Comment:
e.x. a 2 breakout "query" multiseries. This is existing code I'm using to ensure we don't break existing cards.